File: solutions/cpp/lasagna-master/1/lasagna_master.cpp
#include "lasagna_master.h"
#include <iterator>
#include <vector>

namespace lasagna_master {

int preparationTime(std::vector<std::string> &layers, int times) {
  return layers.size() * times;
}

const amount quantities(std::vector<std::string> &layers) {
  auto ret = amount{};
  for (auto const &e : layers) {
    if (e == "noodles") {
      ret.noodles += 50;
    } else if (e == "sauce") {
      ret.sauce += 0.2;
    }
  }
  return ret;
}

void addSecretIngredient(std::vector<std::string> &himel,
                         const std::vector<std::string> &frieren) {
  auto i = himel.end() - 1;
  *i = frieren.at(frieren.size() - 1);
}

const std::vector<double> scaleRecipe(const std::vector<double> &portion,
                                      int scale) {
  std::vector<double> ret(portion.size(), 0.0);

  auto ii = ret.begin();
  for (auto i = portion.begin(); i != portion.end(); i = std::next(i)) {
    *ii = *i * (scale / 2.0);
    ii = std::next(ii);
  }
  return ret;
}

void addSecretIngredient(std::vector<std::string> &himel,
                         const std::string &frieren) {
  auto i = himel.end() - 1;
  *i = frieren;
}

} // namespace lasagna_master

File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagna

import "sort"

// TODO: define the 'PreparationTime()' function
func PreparationTime(layers []string, time int) int {
	if time == 0 {
		return 4
	}
	return len(layers) * time
}

// TODO: define the 'Quantities()' function
func Quantities(layers []string) (int, float64) {
	noodle := 0
	sauce := 0.0
	for _, v := range layers {
		if v == "sauce" {
			sauce += 0.2
		} else if v == "noodles" {
			noodle += 50
		}
	}
	return noodle, sauce
}

// TODO: define the 'AddSecretIngredient()' function
func AddSecretIngredient(otherList, myList []string) {
	index := 0
	for index = len(otherList) - 1; index >= 0; index-- {
		if sort.SearchStrings(myList, otherList[index]) == 0 {
			break
		}
	}
	myList[len(otherList)] = otherList[index]
}

// TODO: define the 'ScaleRecipe()' function
func ScaleRecipe(quantities []float64, portion int) []float64 {
	return []float64{0.0}
}
